### PR TITLE
Added support for retrieving present options and modules

### PR DESCRIPTION
**Pull request by jontje**
As per title.

Related to issue https://github.com/ros-industrial/abb_librws/issues/45.

### CODE REVIEW

File: include/abb_librws/rws_common.h [hunk 1]
```diff
@@ -493,6 +518,11 @@ struct SystemConstants
        * \brief State.
        */
       static const std::string STATE;
+
+      /**
+       * \brief Sys.
```
Comment (gavanderhoorn) [inline include/abb_librws/rws_common.h]: Is there really no description of this?

File: include/abb_librws/rws_interface.h [hunk 2]
```diff
@@ -64,19 +64,75 @@ class RWSInterface
...
+    RobotWareOptionInfo(std::string name, std::string description)
+    :
+    name(name),
+    description(description)
+    {}
+
+    /**
+     * \brief The option's name.
+     */
+    std::string name;
```
Comment (gavanderhoorn) [inline include/abb_librws/rws_interface.h]: Please use `_` suffixes on member variables.

```suggestion
    std::string name_;
```

File: include/abb_librws/rws_interface.h [hunk 3]
```diff
@@ -64,19 +64,75 @@ class RWSInterface
...
+  {
+    /**
+     * \brief A constructor.
+     *
+     * \param name for the name of the option.
+     * \param description for the description of the option.
+     */
+    RobotWareOptionInfo(std::string name, std::string description)
+    :
+    name(name),
```
Comment (gavanderhoorn) [inline include/abb_librws/rws_interface.h]:
```suggestion
    name_(name),
```

File: include/abb_librws/rws_interface.h [hunk 4]
```diff
@@ -64,19 +64,75 @@ class RWSInterface
...
+    description(description)
```
Comment (gavanderhoorn) [inline include/abb_librws/rws_interface.h]:
```suggestion
    description_(description)
```

File: include/abb_librws/rws_interface.h [hunk 5]
```diff
@@ -64,19 +64,75 @@ class RWSInterface
...
+
+    /**
+     * \brief The option's name.
+     */
+    std::string name;
+
+    /**
+     * \brief The options's description.
+     */
+    std::string description;
```
Comment (gavanderhoorn) [inline include/abb_librws/rws_interface.h]:
```suggestion
    std::string description_;
```

File: include/abb_librws/rws_interface.h [hunk 6]
```diff
@@ -64,19 +64,75 @@ class RWSInterface
...
+  {
+    /**
+     * \brief A constructor.
+     *
+     * \param name for the name of the module.
+     * \param type for the type of the module.
+     */
+    RAPIDModuleInfo(std::string name, std::string type)
+    :
+    name(name),
```
Comment (gavanderhoorn) [inline include/abb_librws/rws_interface.h]:
```suggestion
    name_(name),
```

File: include/abb_librws/rws_interface.h [hunk 7]
```diff
@@ -64,19 +64,75 @@ class RWSInterface
...
+    type(type)
```
Comment (gavanderhoorn) [inline include/abb_librws/rws_interface.h]:
```suggestion
    type_(type)
```

File: include/abb_librws/rws_interface.h [hunk 8]
```diff
@@ -64,19 +64,75 @@ class RWSInterface
...
+    {}
+
+    /**
+     * \brief The module's name.
+     */
+    std::string name;
```
Comment (gavanderhoorn) [inline include/abb_librws/rws_interface.h]:
```suggestion
    std::string name_;
```

File: include/abb_librws/rws_interface.h [hunk 9]
```diff
@@ -64,19 +64,75 @@ class RWSInterface
...
+
+    /**
+     * \brief The module's type.
+     */
+    std::string type;
```
Comment (gavanderhoorn) [inline include/abb_librws/rws_interface.h]:
```suggestion
    std::string type_;
```

File: include/abb_librws/rws_interface.h [hunk 10]
```diff
@@ -204,17 +260,24 @@ class RWSInterface
   /**
    * \brief A method for collecting runtime information of the robot controller.
    *
-   * \return RuntimeInfo container for the runtime information.
+   * \return RuntimeInfo containing the runtime information.
```
Comment (gavanderhoorn) [inline include/abb_librws/rws_interface.h]: Hm. Perhaps personal, but I'd prefer the "Container for .." version here, as it describes *what* `RuntimeInfo` *is*, not what it does.

File: include/abb_librws/rws_interface.h [hunk 11]
```diff
@@ -204,17 +260,24 @@ class RWSInterface
...
    * \brief A method for collecting static information (at least during runtime) of the robot controller.
    *
-   * \return StaticInfo container for the static information (at least during runtime).
+   * \return StaticInfo containing the static information (at least during runtime).
    */
   StaticInfo collectStaticInfo();
-  
+
+  /**
+   * \brief A method for retrieving the RobotWare options present in the active robot controller system.
```
Comment (gavanderhoorn) [inline include/abb_librws/rws_interface.h]:
```suggestion
   * \brief A method for retrieving the RobotWare options present on the active robot controller system.
```

File: include/abb_librws/rws_interface.h [hunk 12]
```diff
@@ -204,17 +260,24 @@ class RWSInterface
...
+   *
+   * \return std::vector<OptionInfo> containing a list of the present RobotWare options.
```
Comment (gavanderhoorn) [inline include/abb_librws/rws_interface.h]: Here also: `std::vector<OptionInfo>` is returned, "a list of the .." (ie: I'd not make this active).

File: src/rws_client.cpp [hunk 13]
```diff
@@ -622,6 +644,11 @@ std::string RWSClient::getLogText(const bool verbose)
   return ss.str();
 }
 
+std::string RWSClient::generateConfigurationPath(const std::string& topic, const std::string& type)
+{
+  return Resources::RW_CFG + "/" + topic + "/" + type;
```
Comment (gavanderhoorn) [inline src/rws_client.cpp]: Should we be using [Poco.URI](https://pocoproject.org/docs/Poco.URI.html) here?

And elsewhere we are dealing with/constructing URIs?

File: src/rws_interface.cpp [hunk 14]
```diff
@@ -81,6 +82,27 @@ RWSInterface::StaticInfo RWSInterface::collectStaticInfo()
...
+std::vector<RWSInterface::RobotWareOptionInfo> RWSInterface::getPresentRobotWareOptions()
+{
+  std::vector<RobotWareOptionInfo> result;
+
+  RWSClient::RWSResult rws_result = rws_client_.getConfigurationInstances(Identifiers::SYS,
+                                                                          Identifiers::PRESENT_OPTIONS);
+
+  std::vector<Poco::XML::Node*> node_list = xmlFindNodes(rws_result.p_xml_document, XMLAttributes::CLASS_CFG_IA_T_LI);
+
+  for (size_t i = 0; i < node_list.size(); i+=2)
```
Comment (gavanderhoorn) [inline src/rws_interface.cpp]: Just noticed this: `i+2`.

Should we make sure that `node_list` actually contains a multiple of 2 items?

File: include/abb_librws/rws_client.h [hunk 15]
```diff
@@ -347,6 +347,16 @@ class RWSClient : public POCOClient
     logout();
   }
 
+  /**
+   * \brief A method for retrieving the configuration instances of a type, belonging to a specific configuration topic.
+   *
+   * \param topic specifying the configuration topic.
+   * \param type specifying the type in the configuration topic.
```
[the remark this answers is not in the document]
Comment (jontje) [reply include/abb_librws/rws_client.h]: I don't agree :) I am assuming that the user is at least a bit familiar with how system configurations are structured in RobotWare systems, and then these descriptions should be enough.

File: include/abb_librws/rws_client.h [hunk 16]
```diff
@@ -408,7 +418,16 @@ class RWSClient : public POCOClient
...
-  
+
+  /**
+   * \brief A method for retrieving information about the RAPID modules of a RAPID task.
+   *
+   * \param task specifying the RAPID task.
+   *
+   * \return RWSResult containing the result.
+   */
+  RWSResult getRAPIDModulesInfo(const std::string task);
```
[the remark this answers is not in the document]
Comment (jontje) [reply include/abb_librws/rws_client.h]: Will do.

File: include/abb_librws/rws_client.h [hunk 17]
```diff
@@ -656,6 +675,16 @@ class RWSClient : public POCOClient
...
 
+  /**
+   * \brief Method for generating a configuration URI path.
+   *
+   * \param topic for the configuration topic.
+   * \param type for the configuration type (belonging to the topic).
+   *
+   * \return std::string containing the path.
+   */
+  std::string generateConfigurationPath(const std::string& topic, const std::string& type);
```
[the remark this answers is not in the document]
Comment (jontje) [reply include/abb_librws/rws_client.h]: Yes, and maybe I should reduce my use of copy-pasting and take more care in the implementation :)

File: include/abb_librws/rws_common.h [hunk 18]
```diff
@@ -353,6 +353,11 @@ struct SystemConstants
      */
     struct XMLAttributes
     {
+      /**
+       * \brief Class & cfg-ia-t-li.
```
[the remark this answers is not in the document]
Comment (jontje) [reply include/abb_librws/rws_common.h]: I agree it's not much of a description.

I didn't find an explanation of what `ia-t` means so I didn't really know what to put there. It's `configuration-something-something-list-item` which is the value of the `class` XML attribute.

File: include/abb_librws/rws_common.h (hunk 1 above)
Comment (jontje) [reply include/abb_librws/rws_common.h]: Oops, I forgot this one.

It is the `controller` topic in the configurations, which is abbreviated as `sys`.

File: include/abb_librws/rws_interface.h (hunk 2 above)
Comment (jontje) [reply include/abb_librws/rws_interface.h]: I have used the `_` suffix for private member variables only, and for `struct` data members I have mainly followed the [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html).

File: include/abb_librws/rws_interface.h (hunk 10 above)
Comment (jontje) [reply include/abb_librws/rws_interface.h]: That's resonable, I will think a bit about it.

File: include/abb_librws/rws_interface.h (hunk 12 above)
Comment (jontje) [reply include/abb_librws/rws_interface.h]: Ok.

File: src/rws_client.cpp (hunk 13 above)
Comment (jontje) [reply src/rws_client.cpp]: That's a good idea, I will open an issue for it.

File: src/rws_interface.cpp (hunk 14 above)
Comment (jontje) [reply src/rws_interface.cpp]: There's an if case inside the for loop for that, but what I really should do (and I was thinking about it) is to implement a `xmlFindChildNodes(...)` function to get a list of a XML node's the children that for example has a specific XML attribute.

File: include/abb_librws/rws_client.h (hunk 16 above)
[the remark this answers is not in the document]
Comment (jontje) [reply include/abb_librws/rws_client.h]: I will do this in a future PR and not in this one.

File: include/abb_librws/rws_interface.h (hunk 10 above)
Comment (jontje) [reply include/abb_librws/rws_interface.h]: I will consider this in a future PR.

File: include/abb_librws/rws_interface.h (hunk 12 above)
Comment (jontje) [reply include/abb_librws/rws_interface.h]: I will consider this in a future PR.

File: src/rws_client.cpp (hunk 13 above)
Comment (jontje) [reply src/rws_client.cpp]: I have opened issue https://github.com/ros-industrial/abb_librws/issues/48 for this.

**Comment by gavanderhoorn** [comment]
@jontje: feel free to merge. I somehow not got notified of the fact that you had pushed a new commit.

**Comment by jontje** [comment]
No worries @gavanderhoorn, thanks for the review 👍

**Pull request merged**